File: routes/productos.js
const express = require('express');
const db      = require('../db/database');
const { authMiddleware, soloAdmin } = require('../middleware/auth');

const router = express.Router();

// GET /api/productos — público
router.get('/', (req, res) => {
  const { categoria } = req.query;
  const sql = categoria
    ? 'SELECT * FROM productos WHERE activo = 1 AND categoria = ? ORDER BY nombre'
    : 'SELECT * FROM productos WHERE activo = 1 ORDER BY nombre';
  const rows = categoria
    ? db.prepare(sql).all(categoria)
    : db.prepare(sql).all();
  res.json(rows);
});

// GET /api/productos/:id — público
router.get('/:id', (req, res) => {
  const p = db.prepare('SELECT * FROM productos WHERE id = ? AND activo = 1').get(req.params.id);
  if (!p) return res.status(404).json({ error: 'Producto no encontrado' });
  res.json(p);
});

// POST /api/productos — solo admin
router.post('/', authMiddleware, soloAdmin, (req, res) => {
  const { nombre, descripcion, precio, stock, categoria, emoji, imagen } = req.body;
  if (!nombre || !precio) return res.status(400).json({ error: 'Nombre y precio son requeridos' });

  const info = db.prepare(
    'INSERT INTO productos (nombre, descripcion, precio, stock, categoria, emoji, imagen) VALUES (?,?,?,?,?,?,?)'
  ).run(nombre, descripcion || '', precio, stock || 0, categoria || 'general', emoji || '🥜', imagen || null);

  const nuevo = db.prepare('SELECT * FROM productos WHERE id = ?').get(info.lastInsertRowid);
  res.status(201).json(nuevo);
});

// PUT /api/productos/:id — solo admin
router.put('/:id', authMiddleware, soloAdmin, (req, res) => {
  const { nombre, descripcion, precio, stock, categoria, emoji, imagen, activo } = req.body;
  const existe = db.prepare('SELECT id FROM productos WHERE id = ?').get(req.params.id);
  if (!existe) return res.status(404).json({ error: 'Producto no encontrado' });

  db.prepare(`
    UPDATE productos SET
      nombre      = COALESCE(?, nombre),
      descripcion = COALESCE(?, descripcion),
      precio      = COALESCE(?, precio),
      stock       = COALESCE(?, stock),
      categoria   = COALESCE(?, categoria),
      emoji       = COALESCE(?, emoji),
      imagen      = COALESCE(?, imagen),
      activo      = COALESCE(?, activo)
    WHERE id = ?
  `).run(nombre, descripcion, precio, stock, categoria, emoji, imagen, activo, req.params.id);

  res.json(db.prepare('SELECT * FROM productos WHERE id = ?').get(req.params.id));
});

// DELETE /api/productos/:id — solo admin (soft delete)
router.delete('/:id', authMiddleware, soloAdmin, (req, res) => {
  db.prepare('UPDATE productos SET activo = 0 WHERE id = ?').run(req.params.id);
  res.json({ ok: true });
});

// GET /api/productos/admin/todos — solo admin (incluye inactivos)
router.get('/admin/todos', authMiddleware, soloAdmin, (req, res) => {
  res.json(db.prepare('SELECT * FROM productos ORDER BY activo DESC, nombre').all());
});

module.exports = router;

File: backend/apis/productos/productos.py
from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel
from typing import Optional
from db.database import get_conn
from utils.auth.session import verify_session, solo_admin

router = APIRouter(prefix="/productos", tags=["productos"])

class ProductoBody(BaseModel):
    nombre:      str
    descripcion: Optional[str] = ""
    precio:      int
    stock:       Optional[int] = 0
    categoria:   Optional[str] = "general"
    emoji:       Optional[str] = "🥜"
    imagen:      Optional[str] = None

class ProductoUpdate(BaseModel):
    nombre:      Optional[str] = None
    descripcion: Optional[str] = None
    precio:      Optional[int] = None
    stock:       Optional[int] = None
    categoria:   Optional[str] = None
    emoji:       Optional[str] = None
    imagen:      Optional[str] = None
    activo:      Optional[int] = None

@router.get("")
def listar_productos(categoria: Optional[str] = Query(None)):
    conn = get_conn()
    if categoria:
        rows = conn.execute(
            "SELECT * FROM productos WHERE activo=1 AND categoria=? ORDER BY nombre", (categoria,)
        ).fetchall()
    else:
        rows = conn.execute(
            "SELECT * FROM productos WHERE activo=1 ORDER BY nombre"
        ).fetchall()
    conn.close()
    return [dict(r) for r in rows]

@router.get("/admin/todos")
def listar_todos(user: dict = Depends(solo_admin)):
    conn = get_conn()
    rows = conn.execute("SELECT * FROM productos ORDER BY activo DESC, nombre").fetchall()
    conn.close()
    return [dict(r) for r in rows]

@router.get("/{producto_id}")
def obtener_producto(producto_id: int):
    conn = get_conn()
    row = conn.execute("SELECT * FROM productos WHERE id=? AND activo=1", (producto_id,)).fetchone()
    conn.close()
    if not row:
        raise HTTPException(status_code=404, detail="Producto no encontrado")
    return dict(row)

@router.post("", status_code=201)
def crear_producto(body: ProductoBody, user: dict = Depends(solo_admin)):
    conn = get_conn()
    cur = conn.execute(
        "INSERT INTO productos (nombre,descripcion,precio,stock,categoria,emoji,imagen) VALUES (?,?,?,?,?,?,?)",
        (body.nombre, body.descripcion, body.precio, body.stock, body.categoria, body.emoji, body.imagen),
    )
    conn.commit()
    nuevo = dict(conn.execute("SELECT * FROM productos WHERE id=?", (cur.lastrowid,)).fetchone())
    conn.close()
    return nuevo

@router.put("/{producto_id}")
def actualizar_producto(producto_id: int, body: ProductoUpdate, user: dict = Depends(solo_admin)):
    conn = get_conn()
    existe = conn.execute("SELECT id FROM productos WHERE id=?", (producto_id,)).fetchone()
    if not existe:
        conn.close()
        raise HTTPException(status_code=404, detail="Producto no encontrado")
    campos = {k: v for k, v in body.model_dump().items() if v is not None}
    if campos:
        sets = ", ".join(f"{k}=?" for k in campos)
        conn.execute(f"UPDATE productos SET {sets} WHERE id=?", (*campos.values(), producto_id))
        conn.commit()
    updated = dict(conn.execute("SELECT * FROM productos WHERE id=?", (producto_id,)).fetchone())
    conn.close()
    return updated

@router.delete("/{producto_id}")
def eliminar_producto(producto_id: int, user: dict = Depends(solo_admin)):
    conn = get_conn()
    conn.execute("UPDATE productos SET activo=0 WHERE id=?", (producto_id,))
    conn.commit()
    conn.close()
    return {"ok": True}
